arch/risc-v/src/common/riscv_percpu.h

```diff
@@ -65,16 +65,20 @@
  * will set up [m/s]scratch to point to the CPUs own area
  */
 
-struct riscv_percpu_s
```

Comment:
I see what is going on here, but I find this a bit confusing. Now items of type riscv_percpu_t are list items or not, depending on whether the tcb  pointer is used or not....

If you like it like this, the re-use of of riscv_percpu_s type name is confusing, even though it is not(?) used outside this header. Maybe use "union riscv_percpu_u" here instead in that case?

Reply:
Apparently I should have used _u suffix according to the coding standard.

> Now items of type riscv_percpu_t are list items or not, depending on whether the tcb pointer is used or no
That is almost the case. riscv_percpu_t are not list items ever, unless they are in the freelist for percpu items. They can never be held as a reference while they are in a list, ever, because the only operation for the list is pop() (and insert() during boot).

Another way to hide this detail would be to split this union into two parts:
- private part, which has the list element and put it into riscv_percpu.c
- public part which is exposed here

I'll not change this now, as this is already in upstream, but I see your point. Maybe I'll clean it up at some point.